tests/ignite/engine/test_engine.py

```diff
@@ -159,10 +274,58 @@ def start_of_iteration_handler(engine):
 
     max_epochs = 3
     engine.add_event_handler(Events.ITERATION_STARTED, start_of_iteration_handler)
-    state = engine.run(data, max_epochs=max_epochs, epoch_length=num_iterations_per_epoch)
+    state = engine.run(data, max_epochs=max_epochs, epoch_length=epoch_length)
     # completes the iteration but doesn't increment counter (this happens just before a new iteration starts)
-    true_value = num_iterations_per_epoch * (max_epochs - 1) + iteration_to_stop % num_iterations_per_epoch
+    true_value = real_epoch_length * (max_epochs - 1) + iteration_to_stop % real_epoch_length
     assert state.iteration == true_value
+    assert state.epoch == max_epochs
+
+
+@pytest.mark.parametrize(
+    "terminate_epoch_event, i",
+    [
+        (Events.GET_BATCH_STARTED(once=12), 12),
+        (Events.GET_BATCH_COMPLETED(once=12), 12),
+        (Events.ITERATION_STARTED(once=14), 14),
+        (Events.ITERATION_COMPLETED(once=14), 14),
+    ],
+)
+def test_terminate_epoch_events_sequence(terminate_epoch_event, i):
+    engine = RecordedEngine(MagicMock(return_value=1))
+    data = range(10)
+    max_epochs = 3
+
+    # TODO: Bug: Events.GET_BATCH_STARTED(once=12) is called twice !
```

Comment:
This will be solved when engine is refactored. Yes?
Because now event filter decides on `engine.iteration` (which is not incremented due to termination), but then decides on the actual count of `GET_BATCH_STARTED`.

Reply:
yes, something like that